File: tests/node/assembler/github-hooks-assembler.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import * as fs from "node:fs";
import * as path from "node:path";
import { tmpdir } from "node:os";
import {
  GithubHooksAssembler,
  GITHUB_HOOK_TEMPLATES,
} from "../../../src/assembler/github-hooks-assembler.js";
import { TemplateEngine } from "../../../src/template-engine.js";
import {
  ProjectConfig,
  ProjectIdentity,
  ArchitectureConfig,
  InterfaceConfig,
  LanguageConfig,
  FrameworkConfig,
} from "../../../src/models.js";

function buildConfig(): ProjectConfig {
  return new ProjectConfig(
    new ProjectIdentity("my-app", "A test application"),
    new ArchitectureConfig("microservice", false, false),
    [new InterfaceConfig("rest")],
    new LanguageConfig("typescript", "5"),
    new FrameworkConfig("commander", "", "npm"),
  );
}

function createGithubHookTemplate(
  resourcesDir: string,
  filename: string,
  content: string = '{"event": "test"}',
): void {
  const dir = path.join(resourcesDir, "github-hooks-templates");
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(path.join(dir, filename), content, "utf-8");
}

describe("GithubHooksAssembler", () => {
  let tmpDir: string;
  let resourcesDir: string;
  let outputDir: string;
  let assembler: GithubHooksAssembler;

  beforeEach(() => {
    tmpDir = fs.mkdtempSync(
      path.join(tmpdir(), "gh-hooks-asm-test-"),
    );
    resourcesDir = path.join(tmpDir, "resources");
    fs.mkdirSync(resourcesDir, { recursive: true });
    outputDir = path.join(tmpDir, "output");
    fs.mkdirSync(outputDir, { recursive: true });
    assembler = new GithubHooksAssembler();
  });

  afterEach(() => {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  });

  describe("GITHUB_HOOK_TEMPLATES constant", () => {
    it("contains3Templates", () => {
      expect(GITHUB_HOOK_TEMPLATES).toHaveLength(3);
      expect(GITHUB_HOOK_TEMPLATES).toContain("post-compile-check.json");
      expect(GITHUB_HOOK_TEMPLATES).toContain("pre-commit-lint.json");
      expect(GITHUB_HOOK_TEMPLATES).toContain("session-context-loader.json");
    });
  });

  describe("assemble", () => {
    it("assemble_validConfig_copies3HookFiles", () => {
      for (const template of GITHUB_HOOK_TEMPLATES) {
        createGithubHookTemplate(resourcesDir, template);
      }
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toHaveLength(3);
      for (const template of GITHUB_HOOK_TEMPLATES) {
        const dest = path.join(outputDir, "hooks", template);
        expect(fs.existsSync(dest)).toBe(true);
      }
    });

    it("assemble_templatesDirMissing_returnsEmpty", () => {
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toEqual([]);
      const hooksDir = path.join(outputDir, "hooks");
      expect(fs.existsSync(hooksDir)).toBe(false);
    });

    it("assemble_individualTemplateFileMissing_skipsIt", () => {
      createGithubHookTemplate(resourcesDir, "post-compile-check.json");
      createGithubHookTemplate(resourcesDir, "pre-commit-lint.json");
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toHaveLength(2);
      const missing = path.join(
        outputDir, "hooks", "session-context-loader.json",
      );
      expect(fs.existsSync(missing)).toBe(false);
    });

    it("assemble_createsHooksDirectory_whenNotPreExisting", () => {
      for (const template of GITHUB_HOOK_TEMPLATES) {
        createGithubHookTemplate(resourcesDir, template);
      }
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const hooksDir = path.join(outputDir, "hooks");
      expect(fs.existsSync(hooksDir)).toBe(true);
      expect(fs.statSync(hooksDir).isDirectory()).toBe(true);
    });

    it("assemble_copiedContentMatchesSource", () => {
      const content = '{"event":"post-compile","active":true}';
      createGithubHookTemplate(
        resourcesDir, "post-compile-check.json", content,
      );
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      const copiedContent = fs.readFileSync(result[0]!, "utf-8");
      expect(copiedContent).toBe(content);
    });

    it("assemble_engineParameterNotUsed", () => {
      const content = '{"name": "{project_name}"}';
      createGithubHookTemplate(
        resourcesDir, "post-compile-check.json", content,
      );
      const config = buildConfig();
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      const copiedContent = fs.readFileSync(result[0]!, "utf-8");
      expect(copiedContent).toContain("{project_name}");
    });
  });
});

File: java/src/test/java/dev/iadev/application/assembler/GithubHooksAssemblerTest.java
package dev.iadev.assembler;

import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests for GithubHooksAssembler -- the twelfth
 * assembler in the pipeline, copying hook JSON files
 * verbatim to .github/hooks/.
 */
@DisplayName("GithubHooksAssembler")
class GithubHooksAssemblerTest {

    @Nested
    @DisplayName("implements Assembler interface")
    class ImplementsAssembler {

        @Test
        @DisplayName("is instance of Assembler")
        void instanceOf_whenCreated_implementsAssemblerInterface() {
            GithubHooksAssembler assembler =
                    new GithubHooksAssembler();

            assertThat(assembler)
                    .isInstanceOf(Assembler.class);
        }
    }

    @Nested
    @DisplayName("GITHUB_HOOK_TEMPLATES constant")
    class HookTemplatesConstant {

        @Test
        @DisplayName("contains exactly 3 templates")
        void assemble_whenCalled_containsThreeTemplates() {
            assertThat(
                    GithubHooksAssembler
                            .GITHUB_HOOK_TEMPLATES)
                    .hasSize(3);
        }

        @Test
        @DisplayName("contains expected template names")
        void assemble_whenCalled_containsExpectedNames() {
            assertThat(
                    GithubHooksAssembler
                            .GITHUB_HOOK_TEMPLATES)
                    .containsExactly(
                            "post-compile-check.json",
                            "pre-commit-lint.json",
                            "session-context-loader.json");
        }
    }

    @Nested
    @DisplayName("assemble — copies hook JSON verbatim")
    class AssembleHooks {

        @Test
        @DisplayName("copies 3 hook JSON files to"
                + " hooks/ directory")
        void assemble_whenCalled_copiesThreeHookFiles(
                @TempDir Path tempDir)
                throws IOException {
            Path resourcesDir = createHooksResources(
                    tempDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            List<String> files = assembler.assemble(
                    config, engine, outputDir);

            assertThat(files).hasSize(3);
        }

        @Test
        @DisplayName("creates hooks/ subdirectory")
        void assemble_whenCalled_createsHooksSubdirectory(
                @TempDir Path tempDir)
                throws IOException {
            Path resourcesDir = createHooksResources(
                    tempDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            assembler.assemble(config, engine, outputDir);

            assertThat(outputDir.resolve("hooks"))
                    .exists()
                    .isDirectory();
        }

        @Test
        @DisplayName("output files are verbatim copies"
                + " of templates")
        void assemble_output_isVerbatimCopy(
                @TempDir Path tempDir)
                throws IOException {
            Path resourcesDir = createHooksResources(
                    tempDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            assembler.assemble(config, engine, outputDir);

            for (String name
                    : GithubHooksAssembler
                            .GITHUB_HOOK_TEMPLATES) {
                Path src = resourcesDir.resolve(
                        "github-hooks-templates/" + name);
                Path dest = outputDir.resolve(
                        "hooks/" + name);
                String srcContent = Files.readString(
                        src, StandardCharsets.UTF_8);
                String destContent = Files.readString(
                        dest, StandardCharsets.UTF_8);
                assertThat(destContent)
                        .as("content of %s", name)
                        .isEqualTo(srcContent);
            }
        }

        @Test
        @DisplayName("does not use template engine"
                + " for rendering")
        void assemble_whenCalled_doesNotUseTemplateEngine(
                @TempDir Path tempDir)
                throws IOException {
            Path resourcesDir = createHooksResources(
                    tempDir);
            // Write a template with Pebble syntax
            Path src = resourcesDir.resolve(
                    "github-hooks-templates/"
                            + "post-compile-check.json");
            String contentWithPebble =
                    "{\"name\": \"{{ project_name }}\"}";
            Files.writeString(src, contentWithPebble,
                    StandardCharsets.UTF_8);

            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            assembler.assemble(config, engine, outputDir);

            Path dest = outputDir.resolve(
                    "hooks/post-compile-check.json");
            String result = Files.readString(
                    dest, StandardCharsets.UTF_8);
            // Pebble syntax preserved (not rendered)
            assertThat(result)
                    .isEqualTo(contentWithPebble);
        }

        @Test
        @DisplayName("result paths end with .json")
        void assemble_whenCalled_resultPathsEndWithJson(
                @TempDir Path tempDir)
                throws IOException {
            Path resourcesDir = createHooksResources(
                    tempDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            List<String> files = assembler.assemble(
                    config, engine, outputDir);

            assertThat(files)
                    .allMatch(f -> f.endsWith(".json"));
        }
    }

    @Nested
    @DisplayName("assemble — graceful no-op")
    class GracefulNoOp {

        @Test
        @DisplayName("returns empty list when templates"
                + " directory absent")
        void assemble_whenCalled_returnsEmptyWhenTemplatesDirAbsent(
                @TempDir Path tempDir) {
            Path resourcesDir =
                    tempDir.resolve("nonexistent");
            Path outputDir = tempDir.resolve("output");

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(resourcesDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            List<String> files = assembler.assemble(
                    config, engine, outputDir);

            assertThat(files).isEmpty();
        }

        @Test
        @DisplayName("skips individual missing template")
        void assemble_whenCalled_skipsIndividualMissingTemplate(
                @TempDir Path tempDir)
                throws IOException {
            Path hooksDir = tempDir.resolve(
                    "github-hooks-templates");
            Files.createDirectories(hooksDir);
            // Only create 2 of 3 templates
            Files.writeString(
                    hooksDir.resolve(
                            "post-compile-check.json"),
                    "{}",
                    StandardCharsets.UTF_8);
            Files.writeString(
                    hooksDir.resolve(
                            "pre-commit-lint.json"),
                    "{}",
                    StandardCharsets.UTF_8);
            // session-context-loader.json missing

            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler(tempDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            List<String> files = assembler.assemble(
                    config, engine, outputDir);

            assertThat(files).hasSize(2);
        }
    }

    @Nested
    @DisplayName("assemble — classpath resources")
    class ClasspathResources {

        @Test
        @DisplayName("generates hooks from classpath"
                + " templates")
        void assemble_whenCalled_generatesFromClasspath(
                @TempDir Path tempDir) {
            Path outputDir = tempDir.resolve("output");

            GithubHooksAssembler assembler =
                    new GithubHooksAssembler();
            ProjectConfig config =
                    TestConfigBuilder.minimal();
            TemplateEngine engine = new TemplateEngine();

            List<String> files = assembler.assemble(
                    config, engine, outputDir);

            assertThat(files).hasSize(3);
            assertThat(files)
                    .allMatch(f -> f.endsWith(".json"));
        }
    }

    /**
     * Creates a temporary resources directory with 3
     * hook template JSON files.
     */
    private static Path createHooksResources(Path tempDir)
            throws IOException {
        Path hooksDir = tempDir.resolve(
                "github-hooks-templates");
        Files.createDirectories(hooksDir);
        for (String name
                : GithubHooksAssembler
                        .GITHUB_HOOK_TEMPLATES) {
            Files.writeString(
                    hooksDir.resolve(name),
                    "{\"hook\": \"" + name + "\"}",
                    StandardCharsets.UTF_8);
        }
        return tempDir;
    }
}
